### DISCUSSION TITLE
How to create irregular width fonts?

### DISCUSSION THREAD

**Discussion by q962** [ryanoasis/nerd-fonts#987] [Q&A]
I want to use in GTK.

The purpose is to set all character widths of nerd font to their own displayable width
For example "JetBrains Mono Medium Nerd Font Complete.ttf"

The reason is: while it is possible to display the entire character content by adding spaces, for some characters the spaces are too wide and the right side is too wide. GTK only responds to events within the character width.

For a character:
1: Displayable area width=700, character standard width=600
2: Displayable area width=700=character standard width

For 1, I had to add space after the symbol to receive mouse events beyond the standard width. Otherwise, events in the remaining 100-width area will not be accepted.
On the other hand, the final width is 600+600=1200, but I only need a width of 800, so the area with a width of 400 is blank. Is there a way to eliminate this? Not through gtk. Eliminating this whitespace with gtk is cumbersome.
For 2: I need a width of 800, 700 is less than 800, just center it. this is very simple.

The mono font cannot be used, because when the font is set too large, the whitespace above and below the characters also becomes high. This raises a new question. I can't get rid of it easily by hiding it.

---

So how to solve the problem?

I've also tried setting the space character to 100 for finer tuning.

**Comment by Finii** [comment]
I am not quite shure what you need, but I assume the solution would be a `Nerd Font Proportional`. That font is not monospaced, the letters have all the normal width that they always have (nothing new here), but the width of the symbols is ... well, exactly the width of the symbols. This is usually what you need in 'proportional' contexts (and I assume, GTK is one of them).

See for example the examples here https://github.com/ryanoasis/nerd-fonts/discussions/969#discussioncomment-3881269
What I assume you need is labeled 'partially monospaced'.

Unfortunately we have those fonts not (yet). Can you self-patch? There **is** an option for `font-patcher` that creates such fonts: `--variable-width-glyphs`. If you cant but want to try that approach I can patch something for you.

**Reply by q962** [reply]
I made a mistake. https://github.com/ryanoasis/nerd-fonts/discussions/969 is the discussion I initiated. It seems that I am in a hurry and have not read your reply completely.

It looks like "Proportional" is what I want.

I noticed the '--variable-width-glyphs' parameter, but I don't know how to use it.

**Reply by Finii** [reply]
That parameter is only useful when you patch the font(s) yourself. We do not have 'Proportional' fonts prepatched (yet).

### 1 - direct

You just throw that parameter in on the command line, like:
```
$ fontforge -script font-patcher JetBrainsMono-Regular.ttf --complete --variable-width-glyphs
```

### 2 - docker

Or with docker (the directories are just examples)
```
$ mkdir ~/unpatched
$ cp JetBrainsMono-Regular.ttf ~/unpatched
$ mkdir ~/patched
$ docker run --rm -v ~/unpatched:/in -v ~/patched:/out nerdfonts/patcher --complete --variable-width-glyphs
```

----

### 1

For the first (direct) solution you need to have `fontforge` installed on your system.
To get a working `font-patcher` script you might want to download and extract [`FontPatcher.zip`](https://github.com/ryanoasis/nerd-fonts/releases/download/v2.2.2/FontPatcher.zip).
(Unless you already have the nerd-fonts repo cloned already, then you can use that directly.)

### 2

For the second solution you need to have `docker-ce` installed, see also https://docs.docker.com/engine/install/
With docker you need to specify an input and output directory, and all fonts in the input directory will be patched; you can not specify individual font files.
The docker container is not automatically updated once you pulled it, so you need to do a `docker pull nerdfonts/patcher` from time to time.

If you are lazy like me, you need to `sudo` all docker commands ;-D, otherwise you need to set it up properly.

For this you do not need to have `fontforge` installed, as the container contains an up-to-date version of it.

_Edit: Add link to FontPatcher.zip_

**Reply by q962** [reply]
failed

```sh
                                                                                                                                                                           
# demo @ 172.17.0.1 in ~/Downloads [11:38:14] C:130                                                                                                                                                                                           
$ echo $PWD                                                                                                                                                                                                                                   
/home/demo/Downloads                                                                                                                                                                                                                          
                                                                                                                                                                                                                                              
# demo @ 172.17.0.1 in ~/Downloads [11:38:16]                                                                                                                                                                                                 
$ ls unpatched                                                                                                                                                                                                                                
'JetBrains Mono Regular Nerd Font Complete.ttf'                                                                                                                                                                                               
                                                                                                                                                                                                                                              
# demo @ 172.17.0.1 in ~/Downloads [11:38:28]                                                                                                                                                                                                 
$ ls patched                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                              
# demo @ 172.17.0.1 in ~/Downloads [11:38:32]                                                                                                                                                                                                 
$ sudo dcoker                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                              
# demo @ 172.17.0.1 in ~/Downloads [11:38:38] C:130                                                                                                                                                                                           
$ sudo docker run --rm -v $PWD/unpatched:/in -v $PWD/patched:/out nerdfonts/patcher --complete --variable-width-glyphs                                                                                                                       
Running with options:                                                                                                                                                                                                                         
 --complete --variable-width-glyphs                                                                                                                                                                                                          
                                                                                                                                                                                                                                              
# demo @ 172.17.0.1 in ~/Downloads [11:39:06] C:1          
$ ls patched                                                                                                           

# demo @ 172.17.0.1 in ~/Downloads [11:39:10]              
$ ls unpatched                                             
'JetBrains Mono Regular Nerd Font Complete.ttf'            

```

The image was just downloaded

**Reply by Finii** [reply]
Hmm, somehow the docker container entrypoint script starts, but it can not execute `fontforge` (inside the container).

Here is how it looks like on my machine:

![image](https://user-images.githubusercontent.com/16012374/201855329-2800777d-482d-40e3-8257-ac8115a89ef0.png)

![image](https://user-images.githubusercontent.com/16012374/201855623-7847391e-5546-401a-992c-9743af86292b.png)

_(Notes: You should use the unpatched font as input. No need to create the out-dir first.)_

I used this image
```
$ sudo docker images --digests
REPOSITORY          TAG       DIGEST                                                                    IMAGE ID       CREATED         SIZE
nerdfonts/patcher   latest    sha256:47c7131b7593874c248d90f4269929bdcc015bcd179f6d707264bfa079c7b23d   c3af2ed52eb8   2 weeks ago     90.9MB
hello-world         latest    sha256:7d246653d0511db2a6b2e0436cfd0e52ac8c066000264b3ce63331ac66dca625   feb5d9fea6a5   13 months ago   13.3kB
```

Which platform are you on? I did not test the arm64 container.

**Reply by q962** [reply]
/In/out has no access permission, which may be related to selinux. The '-- privileged=true' parameter is required for run

**Reply by q962** [reply]
Cool, it achieved the effect I needed.
However, I recommend outputting the 'font-family' of the new font

I don't know how to get the information of the font file file via the command line.
I was lucky enough to guess fc-query can be used to query information about a given font file, and I guessed right.

**Reply by Finii** [reply]
Thanks for sharing the `--privileged=true` hint :+1: 

Regarding outputting the Font Family... that is a quick change, never deemed that useful.
And then, there are two (slightly different) Family names in most fonts. Maybe you want to raise an Issue with this as suggestion, so that it is not forgotten?

**Reply by q962** [reply]
Well, I now know that fc-query can get font-family enough
